internal/server/providers/oidc.go

```diff
@@ -0,0 +1,301 @@
+}
+
+// RefreshAccessToken uses the refresh token to get a new access token if it is expired
+func (o *oidcImplementation) RefreshAccessToken(providerUser *models.ProviderUser) (accessToken string, expiry *time.Time, err error) {
+	tokenSource, err := o.tokenSource(providerUser)
+	if err != nil {
+		return "", nil, fmt.Errorf("ref token source: %w", err)
+	}
+
+	newToken, err := tokenSource.Token() // this refreshes token if needed
+	if err != nil {
+		return "", nil, fmt.Errorf("refresh user token: %w", err)
+	}
+
+	return newToken.AccessToken, &newToken.Expiry, nil
+}
+
+func (o *oidcImplementation) SyncProviderUser(db *gorm.DB, user *models.Identity, provider *models.Provider) error {
+	providerUser, err := data.GetProviderUser(db, provider.ID, user.ID)
+	if err != nil {
+		return err
+	}
+
+	if err := checkRefreshAccessToken(db, providerUser, o); err != nil {
+		return fmt.Errorf("oidc sync failed to check users access token: %w", err)
+	}
+
+	info, err := o.GetUserInfo(providerUser)
+	if err != nil {
+		if errors.Is(err, context.DeadlineExceeded) {
+			return fmt.Errorf("%w: %s", internal.ErrBadGateway, err.Error())
+		}
+		return fmt.Errorf("could not get user info from provider: %w", err)
+	}
+
+	logging.S.Debugf("user synchronized with %q groups from provider (ID: %v)", info.Groups, providerUser.ProviderID)
+
+	providerUser.Groups = info.Groups
+	if err := data.UpdateProviderUser(db, providerUser); err != nil {
+		return fmt.Errorf("update provider user: %w", err)
+	}
+
+	if err := data.AssignIdentityToGroups(db, user, provider, info.Groups); err != nil {
+		return fmt.Errorf("assign identity to groups: %w", err)
+	}
+
+	return nil
+}
+
+// GetUserInfo uses a provider token to call the OpenID Connect UserInfo endpoint,
+// make sure an access token is valid (not expired) before using this
+func (o *oidcImplementation) GetUserInfo(providerUser *models.ProviderUser) (*InfoClaims, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), oidcProviderRequestTimeout)
+	defer cancel()
+
+	tokenSource, err := o.tokenSource(providerUser)
+	if err != nil {
+		return nil, fmt.Errorf("info token source: %w", err)
+	}
+
+	_, provider, err := o.clientConfig(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("user info client: %w", err)
+	}
+
+	info, err := provider.UserInfo(ctx, tokenSource)
+	if err != nil {
+		if errors.Is(err, context.DeadlineExceeded) {
+			return nil, err
+		}
+		return nil, fmt.Errorf("get user info: %w", err)
+	}
+
+	claims := &InfoClaims{}
+	if err := info.Claims(claims); err != nil {
+		return nil, fmt.Errorf("user info claims: %w", err)
+	}
+
+	// in the case of azure a deleted user's info will still resolve
+	// guard against this by validating the info in the response is what we expect
+	if err := claims.validate(); err != nil {
+		return nil, err
+	}
+
+	return claims, nil
+}
+
+// validate checks if the user info response claims have the information we expect
+func (ic *InfoClaims) validate() error {
```

Comment:
Should we use our existing way of validating structs?

Reply:
Made the change to use our standard validator pattern.